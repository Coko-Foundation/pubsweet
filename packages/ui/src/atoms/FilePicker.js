import React, { Component } from 'react'
import PropTypes from 'prop-types'

class FilePicker extends Component {
  static propTypes = {
    /** Status of the FilePicker. When disabled no files can be uploaded. */
    disabled: PropTypes.bool,
    /** Allowed files to be uploaded. */
    allowedFileExtensions: PropTypes.arrayOf(PropTypes.string),
    /** Function called with the uploaded file. */
    onUpload: PropTypes.func.isRequired,
  }

  static defaultProps = {
    disabled: false,
    allowedFileExtensions: [],
  }

  handleUpload = e => {
    const { onUpload } = this.props

    onUpload(e.target.files[0])
    this.fileInput.value = null
  }

  getAllowedTypes = () => {
    const { allowedFileExtensions } = this.props

    if (!allowedFileExtensions) {
      return []
    }

    return allowedFileExtensions.map(ext => `.${ext}`)
  }

  render() {
    const { children, disabled } = this.props
    return (
      <div>
        <input
          accept={this.getAllowedTypes()}
          disabled={disabled}
          onChange={this.handleUpload}
          ref={input => (this.fileInput = input)}
          style={{ display: 'none' }}
          type="file"
        />
        {React.cloneElement(React.Children.only(children), {
          tabIndex: disabled ? undefined : 0,
          onKeyPress: e => {
            e.preventDefault()
            if (['Enter', ' '].includes(e.key)) this.fileInput.click()
          },
          onClick: e => {
            e.preventDefault()
            this.fileInput.click()
          },
        })}
      </div>
    )
  }
}

export default FilePicker
