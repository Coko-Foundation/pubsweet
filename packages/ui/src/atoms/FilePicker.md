Using an `<Icon>` component.

```js
const Icon = require('./Icon').default
;<FilePicker onUpload={file => console.log('picked -> ', file)}>
  <div style={{ width: 40 }}>
    <Icon size={36}>file-plus</Icon>
  </div>
</FilePicker>
```

Only .PNG files allowed.

```js
<FilePicker
  onUpload={file => console.log('picked -> ', file)}
  allowedFileExtensions={['png']}
>
  <div>Ony PNGs</div>
</FilePicker>
```

Disabled.

```js
<FilePicker onUpload={file => console.log('picked -> ', file)} disabled>
  <div>I'm disabled</div>
</FilePicker>
```
