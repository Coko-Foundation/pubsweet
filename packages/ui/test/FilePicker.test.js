import React from 'react'
import { mount, shallow } from 'enzyme'

import FilePicker from '../src/atoms/FilePicker'

const props = {
  disabled: false,
  allowedFileExtensions: [],
  onUpload: file => file,
}

describe('FilePicker', () => {
  it('renders correctly', () => {
    const wrapper = shallow(
      <FilePicker {...props}>
        <div>Upload something</div>
      </FilePicker>,
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('parses the allowed extenstions', () => {
    const wrapper = mount(
      <FilePicker {...props} allowedFileExtensions={['pdf', 'png']}>
        <div>Upload something</div>
      </FilePicker>,
    )

    const parsedExtensions = wrapper.instance().getAllowedTypes()
    expect(parsedExtensions).toEqual(['.pdf', '.png'])
  })

  it('picks file', () => {
    const mockFn = jest.fn()
    const wrapper = mount(
      <FilePicker {...props} onUpload={mockFn}>
        <div>Upload something</div>
      </FilePicker>,
    )
    wrapper.instance().handleUpload({ target: { files: [{ name: 'afile' }] } })
    expect(mockFn.mock.calls).toHaveLength(1)
    expect(wrapper.find('input').props().value).toBeFalsy()
  })

  it('renders correct children', () => {
    const divPicker = shallow(
      <FilePicker {...props}>
        <div>Upload something</div>
      </FilePicker>,
    )

    const spanPicker = shallow(
      <FilePicker {...props}>
        <span>Upload something</span>
      </FilePicker>,
    )

    const CustomComp = ({ label }) => <span>{label}</span>

    const customPicker = shallow(
      <FilePicker {...props}>
        <CustomComp label="pick me" />
      </FilePicker>,
    )

    expect(
      divPicker
        .children()
        .last()
        .type(),
    ).toBe('div')
    expect(
      spanPicker
        .children()
        .last()
        .type(),
    ).toBe('span')
    expect(
      customPicker
        .children()
        .last()
        .type(),
    ).toBe(CustomComp)
  })
})
