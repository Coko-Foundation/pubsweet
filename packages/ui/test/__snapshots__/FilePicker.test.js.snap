// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilePicker renders correctly 1`] = `
<div>
  <input
    accept={Array []}
    disabled={false}
    onChange={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
    type="file"
  />
  <div
    onClick={[Function]}
    onKeyPress={[Function]}
    tabIndex={0}
  >
    Upload something
  </div>
</div>
`;
